### PR TITLE
Some API update messages

### PR DESCRIPTION
**Pull request by phansys**
## Subject
Some API update messages.

I am targeting this branch, because these changes respect BC.

Related to #6005.

## Changelog

```markdown
### Deprecated
- Deprecated returning other types than `array` in `AbstractAdmin::getFormGroups()`, `AbstractAdmin::getFormTabs()`, `AbstractAdmin::getShowTabs()`, `AbstractAdmin::getShowGroups()`.
```

## To do
- [x] Add deprecation messages;
- [x] Add the changelog entries.
- [x] Update the tests;

### CODE REVIEW

File: src/Admin/AbstractAdmin.php [hunk 1]
```diff
@@ -566,12 +573,7 @@ abstract class AbstractAdmin implements AdminInterface, DomainObjectInterface, A
      */
     private $filterPersister;
 
-    /**
-     * @param string $code
-     * @param string $class
-     * @param string $baseControllerName
-     */
-    public function __construct($code, $class, $baseControllerName)
+    public function __construct(string $code, string $class, string $baseControllerName)
```
Comment (VincentLanglet) [inline src/Admin/AbstractAdmin.php:576]: It's a BC-break

File: src/Menu/Matcher/Voter/AdminVoter.php [hunk 2]
```diff
@@ -36,7 +36,7 @@ class AdminVoter implements VoterInterface
     /**
      * @var Request
```
Comment (VincentLanglet) [inline src/Menu/Matcher/Voter/AdminVoter.php:37]: Shouldn't the typehint be Request|null ?

File: src/Admin/AbstractAdmin.php (hunk 1 above)
Comment (phansys) [reply src/Admin/AbstractAdmin.php:576]: I guess no. At least, there are no test cases covering this method with other types.

Comment (core23) [reply src/Admin/AbstractAdmin.php:576]: > It's a BC-break

👍 This is a break, some one could use null or other types

File: src/Admin/AbstractAdmin.php [hunk 3]
```diff
@@ -1567,6 +1567,13 @@ public function getMaxPageLinks()
 
     public function getFormGroups()
     {
+        if (!\is_array($this->formGroups) && 'sonata_deprecation_mute' !== (\func_get_args()[0] ?? null)) {
```
Comment (core23) [inline src/Admin/AbstractAdmin.php:1570]: This is right, but how could someone fix this break?

Comment (phansys) [reply src/Admin/AbstractAdmin.php:1570]: I think the simplest option is this:
https://github.com/sonata-project/SonataAdminBundle/blob/7abd1b4e345dcc540cf743bf03bdec744eaf9078/tests/Form/FormMapperTest.php#L66-L68

Other option could be to add some configuration option to tell our codebase it must behave according the relaxed and deprecated API or the new narrower one.

Finally, we could just omit these deprecations and expose these changes in the UPGRADE document.

File: src/Menu/Matcher/Voter/AdminVoter.php (hunk 2 above)
Comment (phansys) [reply src/Menu/Matcher/Voter/AdminVoter.php:37]: Maybe. Let me do some tests.

File: src/Admin/AbstractAdmin.php (hunk 1 above)
Comment (phansys) [reply src/Admin/AbstractAdmin.php:576]: Ok. I'll be adding deprecation notices if other types are used instead.
IMO. the only argument that should accept `string|null` and could be optional is `$baseControllerName`, which is the only one associated to a property with a setter. Do you agree?

File: src/Menu/Matcher/Voter/AdminVoter.php (hunk 2 above)
Comment (phansys) [reply src/Menu/Matcher/Voter/AdminVoter.php]: Done.

File: src/Admin/AbstractAdmin.php (hunk 1 above)
Comment (phansys) [reply src/Admin/AbstractAdmin.php]: Deprecations were added. Please let me know if they seem fine for you.

**Comment by greg0ire** [comment]
Thanks @phansys !

**Pull request merged**